modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/GlobalSnapshotConsensusFunctions.scala

```diff
@@ -112,7 +120,25 @@ object GlobalSnapshotConsensusFunctions {
         updatedLastTxRefs = lastGS.info.lastTxRefs ++ acceptanceResult.contextUpdate.lastTxRefs
         updatedBalances = lastGS.info.balances ++ acceptanceResult.contextUpdate.balances
 
-        rewards = Rewards.calculateRewards(lastGS.proofs.map(_.id))
+        (updatedBalancesWithRewards, rewardTxs) <- trigger match {
+          case EventTrigger => (updatedBalances, SortedSet.empty[RewardTransaction]).pure[F]
+          case TimeTrigger =>
+            rewards
+              .calculateRewards(currentTimeTriggerOrdinal, lastGS.proofs.map(_.id))
+              .flatMap {
+                case (distribution, rewardTxs) =>
+                  Async[F].fromEither {
+                    (updatedBalances.keySet ++ distribution.keySet).toList.traverse { address =>
+                      ((updatedBalances.get(address), distribution.get(address)) match {
+                        case (Some(balance), Some(reward)) => balance.plus(reward)
+                        case (Some(balance), None)         => balance.asRight[BalanceArithmeticError]
+                        case (_, Some(reward))             => Balance.fromAmount(reward).asRight[BalanceArithmeticError]
+                        case _                             => Balance.empty.asRight[BalanceArithmeticError]
+                      }).map((address, _))
+                    }.map(SortedMap.from(_)).map((_, rewardTxs))
```

Comment:
> into another (non-effectful) function

It has to be effectful due to effectful nature of `plus` method. Unless you burn but I am not sure if we should make such decisions right now as it changes the feature parity.

Reply:
It cannot stay this way because it's stopping the consensus. If you need to consult it with someone, please do, but not handling such a case is an error.